### PR TITLE
fix: Fix the race condition in decay average

### PR DESCRIPTION
**Pull request by mutianf**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/java-bigtable/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> ☕️

### CODE REVIEW

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java [hunk 1]
```diff
@@ -72,45 +75,60 @@ boolean setLastAdjustedTimestampMs(long last, long now) {
...
     }
 
-    private double getAlpha(long elapsedSecond) {
-      return Math.exp(-decayConstant * elapsedSecond);
+    private synchronized void updateStartTime(long timestampMs) {
+      long timestampSecond = TimeUnit.MILLISECONDS.toSeconds(timestampMs);
+      long elapsed = timestampSecond - lastUpdateTimeInSecond.get();
+      if (elapsed > UPDATE_START_TIME_THRESHOLD_SECOND) {
+        double decay = getWeight(-elapsed);
+        mean *= decay;
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java:127]: nit. the negative exponent is easy to miss, it might be easier to drop the negative exponent and change this to a division.

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java [hunk 2]
```diff
@@ -72,45 +75,60 @@ boolean setLastAdjustedTimestampMs(long last, long now) {
...
+      return mean * decay;
+    }
+
+    private double getWeight(long elapsedSecond) {
+      return Math.exp(decayConstant * elapsedSecond);
     }
 
-    private double getAlpha(long elapsedSecond) {
-      return Math.exp(-decayConstant * elapsedSecond);
+    private synchronized void updateStartTime(long timestampMs) {
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java:122]: the timestampMs & timestampSeconds is a bit hard to follow. In all cases you end up with the epoch. I think you can improve by using org.threeten.bp.Instant and naming the parameter `now`. Then in you calculations just call `Instant#getEpochSecond()` or `Duration.between(lastUpdate, now).compareTo(UPDATE_START_TIME_THRESHOLD_SECOND) > 0`

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java [hunk 3]
```diff
@@ -72,45 +75,60 @@ boolean setLastAdjustedTimestampMs(long last, long now) {
...
+      // Using weighted count in case the sum overflows.
+      mean =
+          mean * (weightedCount / (weightedCount + weight))
+              + weight * value / (weightedCount + weight);
+      weightedCount += weight;
+
+      // Set last update time so when we're getting the mean we can calculate the decay based on the
+      // last time the mean was updated.
+      if (now > lastUpdateTimeInSecond.get()) {
         lastUpdateTimeInSecond.set(now);
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java:107]: why is this still necessary? isnt this done in updateStartTime now?

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java [hunk 4]
```diff
@@ -72,45 +75,60 @@ boolean setLastAdjustedTimestampMs(long last, long now) {
     private double mean;
     private double weightedCount;
     private AtomicLong lastUpdateTimeInSecond;
+    private long startTimeSecond;
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java:78]: why do we need to track both lastUpdateTimeInSecond & startTimeSecond? don't we just need to track when the last time decay changed?

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java [hunk 5]
```diff
@@ -72,45 +75,60 @@ boolean setLastAdjustedTimestampMs(long last, long now) {
...
       }
     }
 
     double getMean(long timestampMs) {
-      long timestampSecs = TimeUnit.MILLISECONDS.toSeconds(timestampMs);
-      long elapsed = timestampSecs - lastUpdateTimeInSecond.get();
-      return mean * getAlpha(Math.max(0, elapsed));
+      long timestampSecond = TimeUnit.MILLISECONDS.toSeconds(timestampMs);
+      long elapsed = timestampSecond - lastUpdateTimeInSecond.get();
+      double decay = getWeight(Math.min(0, -elapsed));
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java:114]: same as below, I think it would be better to use division instead of negative exponent

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java [hunk 6]
```diff
@@ -72,45 +75,60 @@ boolean setLastAdjustedTimestampMs(long last, long now) {
...
-      Preconditions.checkArgument(
-          now >= lastUpdateTimeInSecond.get(), "can't update an event in the past");
-      if (lastUpdateTimeInSecond.get() == 0) {
+      long elapsed = now - startTimeSecond;
+      double weight = getWeight(elapsed);
+      // Exponential moving average = weightedSum / weightedCount, where
+      // weightedSum(n) = weight(n) * value(n) + weightedSum(n - 1)
+      // weightedCount(n) = weight(n) + weightedCount(n - 1), where weight(n) grows exponentially
+      // over elapsed time.
+      // Using weighted count in case the sum overflows.
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java:98]: Can we bubble this update to class level doc?

**Review by igorbernstein2 — CHANGES_REQUESTED** [review]

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java (hunk 3 above)
Comment (mutianf) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java:107]: Hmm, so if getMean is called after like 5 minutes of update (before we need to update start time) the average will be decayed?

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java (hunk 2 above)
Comment (mutianf) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java]: passing in a timestamp is just for testing, moved this up to be converted when calling this method

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java [hunk 7]
```diff
@@ -15,48 +15,52 @@
...
+  private static final double DEFAULT_DECAY_CONSTANT = 0.015;
+  // Update start time every 15 minutes so the values won't be infinite
+  private static final long UPDATE_START_TIME_THRESHOLD_SECOND = TimeUnit.MINUTES.toSeconds(15);
 
   private AtomicLong lastAdjustedTimestampMs;
   private DecayingAverage meanLatency;
 
   DynamicFlowControlStats() {
-    this(DEFAULT_DECAY_CONSTANT);
+    this(DEFAULT_DECAY_CONSTANT, null);
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java:45]: NanoClock.getDefaultClock()

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java [hunk 8]
```diff
@@ -15,48 +15,52 @@
...
   }
 
-  DynamicFlowControlStats(double decayConstant) {
-    this.lastAdjustedTimestampMs = new AtomicLong(0);
-    this.meanLatency = new DecayingAverage(decayConstant);
+  DynamicFlowControlStats(ApiClock clock) {
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java:48]: I would drop this ctor

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java [hunk 9]
```diff
@@ -71,46 +75,58 @@ boolean setLastAdjustedTimestampMs(long last, long now) {
...
 
-    DecayingAverage(double decayConstant) {
+    DecayingAverage(double decayConstant, ApiClock clock) {
       this.decayConstant = decayConstant;
       this.mean = 0.0;
       this.weightedCount = 0.0;
-      this.lastUpdateTimeInSecond = new AtomicLong(0);
+      this.clock = clock;
+      this.startTimeSecond =
+          clock == null
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java:87]: I think you should make the clock be non-null here.

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java [hunk 10]
```diff
@@ -71,46 +75,58 @@ boolean setLastAdjustedTimestampMs(long last, long now) {
...
+    double getMean() {
+      return mean;
     }
 
-    double getMean(long timestampMs) {
-      long timestampSecs = TimeUnit.MILLISECONDS.toSeconds(timestampMs);
-      long elapsed = timestampSecs - lastUpdateTimeInSecond.get();
-      return mean * getAlpha(Math.max(0, elapsed));
+    private long getCurrentTimeInSecond() {
+      return clock == null
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java:113]: I think you should fix the clock to be non-null and avoid uncertainty here

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java [hunk 11]
```diff
@@ -71,46 +75,58 @@ boolean setLastAdjustedTimestampMs(long last, long now) {
...
-            mean * ((weightedCount * alpha) / (weightedCount * alpha + 1))
-                + value / (weightedCount * alpha + 1);
-        weightedCount = weightedCount * alpha + 1;
-      }
+    synchronized void update(long value) {
+      // Decay mean and weightedCount if now - startTime > threshold, so weight won't be infinite
+      long now = getCurrentTimeInSecond();
+      double decay = getDecay(now);
+      mean /= decay;
+      weightedCount /= decay;
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java:97]: I think you should inline this in getWeight/getDecay

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java [hunk 12]
```diff
@@ -71,46 +75,58 @@ boolean setLastAdjustedTimestampMs(long last, long now) {
...
+          : TimeUnit.MILLISECONDS.toSeconds(clock.millisTime());
     }
 
-    private double getAlpha(long elapsedSecond) {
-      return Math.exp(-decayConstant * elapsedSecond);
+    private double getWeight(long elapsedSecond) {
+      return Math.exp(decayConstant * elapsedSecond);
+    }
+
+    private synchronized double getDecay(long now) {
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java:122]: I would merge these

**Review by igorbernstein2 — CHANGES_REQUESTED** [review]

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java [hunk 13]
```diff
@@ -71,46 +71,45 @@ boolean setLastAdjustedTimestampMs(long last, long now) {
     private double decayConstant;
     private double mean;
     private double weightedCount;
-    private AtomicLong lastUpdateTimeInSecond;
+    private long startTimeSecond;
+    private ApiClock clock;
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java:75]: final

File: google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStatsTest.java [hunk 14]
```diff
@@ -27,40 +27,39 @@
...
+import org.threeten.bp.Instant;
 
 @RunWith(JUnit4.class)
 public class DynamicFlowControlStatsTest {
 
   @Test
   public void testUpdate() {
-    DynamicFlowControlStats stats = new DynamicFlowControlStats();
-    long now = System.currentTimeMillis();
+    FakeApiClock clock = new FakeApiClock(Instant.now().toEpochMilli());
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStatsTest.java:37]: I think a mock would be better for this

```java
@Rule MockitoRule mockitoRule = MockitoJUnitRule.create();
@Mock ApiClock clock;

@Test
public void myTest() {
  Mockito.when(clock.getMillis()).thenReturn(1234);
}
```

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java [hunk 15]
```diff
@@ -15,48 +15,48 @@
...
 
-  private static final double DEFAULT_DECAY_CONSTANT = 0.015; // Biased to the past 5 minutes
+  // Biased to the past 5 minutes (300 seconds), e^(-decay_constant * 300) = 0.01, decay_constant ~=
+  // 0.015
+  private static final double DEFAULT_DECAY_CONSTANT = 0.015;
+  // Update start time every 15 minutes so the values won't be infinite
+  private static final long UPDATE_START_TIME_THRESHOLD_SECOND = TimeUnit.MINUTES.toSeconds(15);
 
   private AtomicLong lastAdjustedTimestampMs;
   private DecayingAverage meanLatency;
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java:42]: final

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java [hunk 16]
```diff
@@ -15,48 +15,48 @@
...
 final class DynamicFlowControlStats {
 
-  private static final double DEFAULT_DECAY_CONSTANT = 0.015; // Biased to the past 5 minutes
+  // Biased to the past 5 minutes (300 seconds), e^(-decay_constant * 300) = 0.01, decay_constant ~=
+  // 0.015
+  private static final double DEFAULT_DECAY_CONSTANT = 0.015;
+  // Update start time every 15 minutes so the values won't be infinite
+  private static final long UPDATE_START_TIME_THRESHOLD_SECOND = TimeUnit.MINUTES.toSeconds(15);
 
   private AtomicLong lastAdjustedTimestampMs;
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java:41]: final

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java [hunk 17]
```diff
@@ -15,48 +15,48 @@
...
   private DecayingAverage meanLatency;
 
   DynamicFlowControlStats() {
-    this(DEFAULT_DECAY_CONSTANT);
+    this(DEFAULT_DECAY_CONSTANT, NanoClock.getDefaultClock());
   }
 
-  DynamicFlowControlStats(double decayConstant) {
+  DynamicFlowControlStats(double decayConstant, ApiClock clock) {
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java:48]: InternalApi("visible for testing");

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java [hunk 18]
```diff
@@ -71,46 +71,45 @@ boolean setLastAdjustedTimestampMs(long last, long now) {
     private double decayConstant;
     private double mean;
     private double weightedCount;
-    private AtomicLong lastUpdateTimeInSecond;
+    private long startTimeSecond;
```
Comment (igorbernstein2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java:74]: I dont think its the startTime anymore. Maybe decayCycleEpoch?

**Review by igorbernstein2 — APPROVED** [review]
lgtm!
